### PR TITLE
Implement P2520R0: `move_iterator<T*>` should be a random access iterator

### LINKED ISSUE #2939: P2520R0 `move_iterator<T*>` Should Be A Random-Access Iterator
[P2520R0](https://wg21.link/P2520R0) `move_iterator<T*>` Should Be A Random-Access Iterator

Feature-test macro (expected):
```cpp
#define __cpp_lib_move_iterator_concept 202207L
```

### PR DESCRIPTION
**Pull request by strega-nil-ms**
Fixes #2939

### CODE REVIEW

File: stl/inc/xutility [hunk 1]
```diff
@@ -3186,11 +3186,26 @@ class move_sentinel;
...
+            return random_access_iterator_tag{};
+        } else if constexpr (bidirectional_iterator<_Iter>) {
+            return bidirectional_iterator_tag{};
+        } else if constexpr (forward_iterator<_Iter>) {
+            return forward_iterator_tag{};
+        } else {
+            return input_iterator_tag{};
+        }
+    }
+    using iterator_concept = decltype(_Get_concept());
```
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/xutility:3202]: Speaking of "unconditional", we should wrap this in `#ifdef __cpp_lib_concepts` to avoid defining `iterator_concept` in the "wrong" language modes.

File: stl/inc/yvals_core.h [hunk 2]
```diff
@@ -1409,6 +1410,7 @@
 #define __cpp_lib_latch                   201907L
 #define __cpp_lib_list_remove_return_type 201806L
 #define __cpp_lib_math_constants          201907L
+#define __cpp_lib_move_iterator_concept   202207L
```
Comment (CaseyCarter) [inline stl/inc/yvals_core.h:1413]: We're missing corresponding changes in `VSO_0157762_feature_test_macros`.

**Review by CaseyCarter — CHANGES_REQUESTED** [review]

File: stl/inc/xutility [hunk 3]
```diff
@@ -3226,6 +3211,19 @@ public:
     using pointer         = _Iter;
 
 #ifdef __cpp_lib_concepts
+    constexpr static auto _Get_iter_concept() {
```
Comment (CaseyCarter) [inline stl/inc/xutility:3214]: Nit: is there a convenient `__cpp_lib_concepts` section in a `private` part of the class we could move this into? We try (but are not completely consistent about) to avoid exposing even `_Ugly` names in `public`.

**Review by CaseyCarter — APPROVED** [review]

File: stl/inc/yvals_core.h (hunk 2 above)
Comment (frederick-vs-ja) [inline stl/inc/yvals_core.h:1413]: Not sure whether to change. Is P2520R0 "Yet another DR against C++20"?

**Comment by CaseyCarter** [comment]
The test runner has found the pertinent libc++ test that we'll need to skip for you.

File: stl/inc/yvals_core.h (hunk 2 above)
Comment (CaseyCarter) [reply stl/inc/yvals_core.h:1413]: Nicole and I discussed this in person. This paper is effectively an admission that artificially depressing the category of `move_iterator` was a failed experiment. There's no value in preserving the undesired behavior in C++20 mode, and we suspect the potential for breakage to be very small given that `move_iterator`s are still input iterators even if they now sometimes satisfy stronger concepts. We therefore decided to implement this change in all pertinent language modes.

Comment (frederick-vs-ja) [reply stl/inc/yvals_core.h:1413]: 😸Thank you for explanation! It sounds to me that this paper is at least a defect report in more generalized meaning.

File: tests/libcxx/skipped_tests.txt [hunk 4]
```diff
@@ -89,6 +89,9 @@ utilities\format\format.functions\vformat.pass.cpp
 # libc++ doesn't implement P2321R2's changes to vector<bool>::reference
 containers\sequences\vector.bool\iterator_concept_conformance.compile.pass.cpp
 
+# libc++ hasn't updated move_iterator for P2520R0
+std/iterators/predef.iterators/move.iterators/move.iterator/types.pass.cpp
```
Comment (CaseyCarter) [inline tests/libcxx/skipped_tests.txt:93]: We have two different files (`expected_results.txt` and `skipped_tests.txt`) describing this test suite for the GitHub and Internal (respectively) test runners. Generally, that means that changes to the two are "mirrored". Changes are never actually identical, however, since (1) this file uses `\` separators and the other `/` separators, (2) paths here are relative to `llvm-project/libcxx/test/std` vs `llvm-project/libcxx/test` for `expected_results.txt`, and (3) this file can only skip tests whereas `expected_results` supports SKIPPED, FAIL, and even PASS (for .fail.cpp tests which don't FAIL) statuses.

TLDR: this line needs to change to:
```suggestion
iterators\predef.iterators\move.iterators\move.iterator\types.pass.cpp
```
and you need to add a section in the same place in `expected_results.txt`:
```c++
# libc++ hasn't updated move_iterator for P2520R0
std/iterators/predef.iterators/move.iterators/move.iterator/types.pass.cpp FAIL
```

**Review by CaseyCarter — CHANGES_REQUESTED** [review]

File: tests/libcxx/skipped_tests.txt (hunk 4 above)
Comment (CaseyCarter) [reply tests/libcxx/skipped_tests.txt:93]: (Remember yesterday when I mentioned how badly we need to port the GitHub test runner into the internal test infra? This is a lot of the motivation =))

**Review by miscco — APPROVED** [review]

**Review by CaseyCarter — APPROVED** [review]

File: stl/inc/xutility [hunk 5]
```diff
@@ -3211,8 +3211,23 @@ public:
     using pointer         = _Iter;
 
 #ifdef __cpp_lib_concepts
-    using iterator_concept = input_iterator_tag;
-    using reference        = iter_rvalue_reference_t<_Iter>;
+private:
+    constexpr static auto _Get_iter_concept() {
```
Comment (StephanTLavavej) [inline stl/inc/xutility:3215]: Style nitpick: We always say `static constexpr`.

File: tests/std/tests/P0896R4_ranges_iterator_machinery/test.cpp [hunk 6]
```diff
@@ -3271,17 +3271,19 @@ struct std::common_type<move_iterator_test::input_iter<true>::rvalue_reference,
...
     STATIC_ASSERT(same_as<move_iterator<simple_bidi_iter<>>::iterator_category, bidirectional_iterator_tag>);
-    STATIC_ASSERT(same_as<move_iterator<simple_forward_iter<>>::iterator_concept, input_iterator_tag>);
+    STATIC_ASSERT(same_as<move_iterator<simple_forward_iter<>>::iterator_concept, forward_iterator_tag>);
     STATIC_ASSERT(same_as<move_iterator<simple_forward_iter<>>::iterator_category, forward_iterator_tag>);
     STATIC_ASSERT(same_as<move_iterator<simple_input_iter>::iterator_concept, input_iterator_tag>);
     STATIC_ASSERT(same_as<move_iterator<simple_input_iter>::iterator_category, input_iterator_tag>);
+    STATIC_ASSERT(same_as<move_iterator<input_iter<true>>::iterator_concept, input_iterator_tag>);
     STATIC_ASSERT(!has_member_iter_category<move_iterator<input_iter<true>>>);
+    STATIC_ASSERT(same_as<move_iterator<input_iter<true>>::iterator_concept, input_iterator_tag>);
     STATIC_ASSERT(!has_member_iter_category<move_iterator<input_iter<false>>>);
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_ranges_iterator_machinery/test.cpp:3287]: The new test line is exactly repeated; it appears that the second one intended to test `input_iter<false>`.

File: tests/libcxx/expected_results.txt [hunk 7]
```diff
@@ -89,6 +89,8 @@ std/utilities/format/format.functions/vformat.pass.cpp FAIL
 # libc++ doesn't implement P2321R2's changes to vector<bool>::reference
 std/containers/sequences/vector.bool/iterator_concept_conformance.compile.pass.cpp FAIL
 
+# libc++ hasn't updated move_iterator for P2520R0
+std/iterators/predef.iterators/move.iterators/move.iterator/types.pass.cpp FAIL
```
Comment (StephanTLavavej) [inline tests/libcxx/expected_results.txt:94]: Nitpick, not worth resetting testing by itself: `skipped_tests.txt` is preserving the 2 newlines that were here, but `expected_results.txt` isn't. This doesn't *seriously* complicate diffing, but we should try to keep these files in sync as much as possible.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
@strega-nil-ms @CaseyCarter I pushed changes for two trivial nitpicks and one copy-pasta, FYI.

**Review by CaseyCarter — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for implementing this C++23 feature and *moving* towards feature completeness! :chart_with_downwards_trend: :joy_cat: :rocket:
